google-cloud-core/src/main/java/com/google/cloud/Policy.java

```diff
@@ -158,64 +174,134 @@ protected Builder(Policy policy) {
      *
      * @throws NullPointerException if the given map is null or contains any null keys or values
      * @throws IllegalArgumentException if any identities in the given map are null
+     * @throws IllegalArgumentException if policy version is equal to 3 or has conditional bindings.
      */
     public final Builder setBindings(Map<Role, Set<Identity>> bindings) {
       checkNotNull(bindings, "The provided map of bindings cannot be null.");
+      checkArgument(
+          !isConditional(this.version, this.bindingsList),
+          "setBindings() is only supported with version 1 policies and non-conditional policies");
       for (Map.Entry<Role, Set<Identity>> binding : bindings.entrySet()) {
         checkNotNull(binding.getKey(), "The role cannot be null.");
         Set<Identity> identities = binding.getValue();
         checkNotNull(identities, "A role cannot be assigned to a null set of identities.");
         checkArgument(!identities.contains(null), "Null identities are not permitted.");
       }
-      this.bindings.clear();
+      // convert into List format
+      this.bindingsList.clear();
       for (Map.Entry<Role, Set<Identity>> binding : bindings.entrySet()) {
-        this.bindings.put(binding.getKey(), new HashSet<>(binding.getValue()));
+        Binding.Builder bindingBuilder = Binding.newBuilder();
+        bindingBuilder.setRole(binding.getKey().getValue());
+        for (Identity identity : binding.getValue()) {
+          bindingBuilder.addMembers(identity.strValue());
+        }
+        this.bindingsList.add(bindingBuilder.build());
+      }
+      return this;
+    }
+
+    /**
+     * Replaces the builder's List of bindings with the given List of Bindings.
+     *
+     * @throws NullPointerException if the given map is null or contains any null keys or values
+     * @throws IllegalArgumentException if any identities in the given map are null
+     */
+    public final Builder setBindings(List<Binding> bindings) {
+      for (Binding binding : bindings) {
+        checkNotNull(binding.getRole(), "The role cannot be null.");
+        List<String> members = binding.getMembers();
+        checkNotNull(members, "A role cannot be assigned to a null set of identities.");
+        checkArgument(!members.contains(null), "Null identities are not permitted.");
```

Comment:
apologies as I'm not super familiar with the code - is there any way to catch this earlier on, like when things get added to `members`? Might be easier/faster than checking for null elements after the fact.

Reply:
It's something that's being done by the existing setBindings(Map<>) overload. Would you prefer I remove it?

Reply:
I think that may make more sense. My concern is every time `contains` is called we're iterating through the list, where we could just throw the exception when an element is added instead.

Reply:
Deleted my previous comment as I misunderstood something - I would add a null check if necessary to wherever you are calling `bindingsList.add` to ensure you're not adding any null elements to the list. That way, by the time you get here, you are sure you don't have any null items and do not need to call `contains` and iterate through the list again.